### PR TITLE
Dockerfile

### LINKED ISSUE #15: Dockerfile
Create a dockerfile that can build an image for running the application. Alpine based jre for java17 looks to be available here now:
https://hub.docker.com/layers/eclipse-temurin/library/eclipse-temurin/17-jre-alpine/images/sha256-839f3208bfc22f17bf57391d5c91d51c627d032d6900a0475228b94e48a8f9b3?context=explore

### PR DESCRIPTION
**Pull request by OscarStjernfeldt**
Implements docker - fixes #15

### CODE REVIEW

**Review by DarkendHall — COMMENTED** [review]
Hard to see if Docker will run it. Is the first part of the Dockerfile really necessary?

**Comment by kappsegla** [comment]
Is maven-assembly-plugin needed?

**Comment by DarkendHall** [comment]
I had problems with dependencies not being included in the .jar file, same as @OscarStjernfeldt and @ChristianLofqvist without it.

**Comment by kappsegla** [comment]
> Hard to see if Docker will run it. Is the first part of the Dockerfile really necessary?

To test this i suggest that you check out this branch and try it out locally on your computer to see that it works as intended. Especially since it's not code that is checked automatically with an action.

**Review by DarkendHall — APPROVED** [review]
Runs just fine when I load it into Docker. Good job 👍

**Review by DarkendHall — CHANGES_REQUESTED** [review]
Just realized the issue was updated with additional requirements, which currently aren't being fulfilled.

Add support for mounting a volume.

**Comment by kappsegla** [comment]
> > Is maven-assembly-plugin needed?
> 
> I had problems with dependencies not being included in the .jar file, same as @OscarStjernfeldt and @ChristianLofqvist without it.

Since the project has a module-info.java file and is a module based project we need to manually copy all dependencies as jar files into the docker image. So at the moment the code seems to have one runtime dependency in the form av log4j-core that might be needed when trying to run the application. So the jar file for that module should be copied into a folder in the docker image and added as --module-path when starting the application instead of using entrypoint java -jar.
What we can do is to have maven put all dependencies that we need into one folder under target and then we can copy that folder into our docker image.

**Comment by DarkendHall** [comment]
> > > Is maven-assembly-plugin needed?
> > 
> > 
> > I had problems with dependencies not being included in the .jar file, same as @OscarStjernfeldt and @ChristianLofqvist without it.
> 
> Since the project has a module-info.java file and is a module based project we need to manually copy all dependencies as jar files into the docker image. So at the moment the code seems to have one runtime dependency in the form av log4j-core that might be needed when trying to run the application. So the jar file for that module should be copied into a folder in the docker image and added as --module-path when starting the application instead of using entrypoint java -jar. What we can do is to have maven put all dependencies that we need into one folder under target and then we can copy that folder into our docker image.

Isn't this what the maven-assembly-plugin does but puts them inside the .jar file instead so you can use java -jar?

**Comment by kappsegla** [comment]
> > > > Is maven-assembly-plugin needed?
> > > 
> > > 
> > > I had problems with dependencies not being included in the .jar file, same as @OscarStjernfeldt and @ChristianLofqvist without it.
> > 
> > 
> > Since the project has a module-info.java file and is a module based project we need to manually copy all dependencies as jar files into the docker image. So at the moment the code seems to have one runtime dependency in the form av log4j-core that might be needed when trying to run the application. So the jar file for that module should be copied into a folder in the docker image and added as --module-path when starting the application instead of using entrypoint java -jar. What we can do is to have maven put all dependencies that we need into one folder under target and then we can copy that folder into our docker image.
> 
> Isn't this what the maven-assembly-plugin does but puts them inside the .jar file instead so you can use java -jar?

Yes but an alternative that is more docker friendly is to use:

<plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-dependency-plugin</artifactId>
                <version>3.1.2</version>
                <executions>
                    <execution>
                        <id>copy</id>
                        <phase>package</phase>
                        <goals>
                            <goal>copy-dependencies</goal>
                        </goals>
                        <configuration>
                            <outputDirectory>
                                ${project.build.directory}/modules
                            </outputDirectory>
                            <includeScope>runtime</includeScope>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
            
            And then in Dockerfile  COPY target/modules /app/modules/
            
            When starting application then we can run ENTRYPOINT [ "java", "--module-path", "/app/modules:...."
            
            That helps in optimizing the docker image so we don't have to copy all the data from dependencies over and over just because our main code changes. Dependency files should preferable be copied first before our own code so it creates it's own layer in the image that doesn't change too often. Has to do with docker image optimization.

**Comment by DarkendHall** [comment]
Tried to build with docker and it worked great, it added a new volume and started up without errors. I don't know if there is a way to test the volume at the moment since we don't have any external files yet.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,10 @@
+FROM maven:3.8.4-openjdk-17-slim as build
+COPY ./ /src
+RUN mvn -f /src/pom.xml clean package
+
+FROM eclipse-temurin:17-jre-alpine
+COPY --from=build /src/target/modules /app/modules
+COPY --from=build /src/target/storm.jar /app/storm.jar
+RUN mkdir /webroot
+VOLUME /webroot
```
Comment (kappsegla) [inline Dockerfile:9]: This isn't needed. 
```suggestion
```

**Review by DarkendHall — APPROVED** [review]

**Comment by LordRekishi** [comment]
I'll take a look at this now.

**Comment by LordRekishi** [comment]
@OscarStjernfeldt what is /webroot supposed to do in Dockerfile?
Also, do we need .dockerignore?

**Comment by OscarStjernfeldt** [comment]
> @OscarStjernfeldt what is /webroot supposed to do in Dockerfile? Also, do we need .dockerignore?

webroot is needed for upcoming mounting. What do we want to ignore with .dockerignore? Please share if you found anything.

**Comment by LordRekishi** [comment]
> > @OscarStjernfeldt what is /webroot supposed to do in Dockerfile? Also, do we need .dockerignore?
> 
> webroot is needed for upcoming mounting. What do we want to ignore with .dockerignore? Please share if you found anything.

Ok, so webroot doesn't do anything at the moment?
We could use the same ignore files as in the project Martin made during our classes.

target/
.github/
.idea/

Should be fine right? The .github is only needed for the project file and not in the docker image since it will be built for release?

**Comment by LordRekishi** [comment]
> I tried adding a .dockerignore on my local copy of this and it didn't affect the size of the image at all, it did however reduce the build time by about 1 second.

Right now size is not a big factor, but in the future when we have more code it will. I am guessing we will implement size saving features later on, like using another base image than the one we are using currently. Like we did in the class practice project.

**Comment by DarkendHall** [comment]
But in the future the time saved from the build process could increase from one second, and then we would need a .dockerignore to reduce it. We could add it and update as we go to reduce build time as much as possible.

**Review by LordRekishi — CHANGES_REQUESTED** [review]
I suggest adding a .dockerignore file with the following lines:

target/
.github/
.idea/

File: .dockerignore [hunk 2]
```diff
@@ -0,0 +1,3 @@
+target/
+.github/
+.idea/
```
Comment (DarkendHall) [inline .dockerignore:3]: No new line at end of file, please add one :)

**Comment by ChristianLofqvist** [comment]
> I suggest adding a .dockerignore file with the following lines:
> 
> target/ .github/ .idea/

Good idea, I added a .dockerignore file with your suggested content.

File: .dockerignore (hunk 2 above)
Comment (kappsegla) [inline .dockerignore]: Add a newline att end of file.
```suggestion
.idea/

```

Comment (ChristianLofqvist) [reply .dockerignore]: My bad, new line added!

**Review by kappsegla — CHANGES_REQUESTED** [review]

File: .dockerignore (hunk 2 above)
Comment (DarkendHall) [reply .dockerignore]: This change was already added

**Review by DarkendHall — APPROVED** [review]

**Review by kappsegla — APPROVED** [review]

**Review by kappsegla — APPROVED** [review]

**Review by LordRekishi — APPROVED** [review]

**Pull request merged**